### PR TITLE
Ignore kernel modules with null addresses

### LINKED ISSUE #157: Issue with some kernel modules
Hello,
I’m trying to run the agent on my server but it crashes during startup.
The problem seems to come from one kernel module not being mapped correctly:

```
Failed to load eBPF tracer: failed to read kernel modules: addresses from modules is zero - check process permissions: 'nls_ucs2_utils 8192 1 cifs, Live 0x0000000000000000' 
```

and indeed, if I have a look at the `/proc/modules` file as root, I can see all modules having a `valid` address and this module `nls_ucs2_utils` having address `0x0000000000000000`.

My understanding is that this check is here for permission validation, could it be possible not to fail if only few modules have address 0x0 but the remaining modules having a `valid` address?

### PR DESCRIPTION
**Pull request by mchataigner**
Fix #157
Only crash the process if all modules have null addresses

### CODE REVIEW

File: proc/proc.go [hunk 1]
```diff
@@ -118,10 +121,12 @@ func GetKernelModules(modulesPath string,
...
-			return nil, fmt.Errorf(
-				"addresses from modules is zero - "+
-					"check process permissions: '%s'", line)
+
+		if address != 0 {
+			atLeastOneValidAddress = true
+		} else {
+			lastInvalidAddressLine = line
+			continue
 		}
```
Comment (fabled) [inline proc/proc.go:130]:
```suggestion
		if address == 0 {
			lastInvalidAddressLine = line
			continue
		}
		atLeastOneValidAddress = true
```
Since the block has `continue`, no need for the else block.

File: proc/proc.go [hunk 2]
```diff
@@ -130,6 +135,13 @@ func GetKernelModules(modulesPath string,
 			Size:    int(size),
 		})
 	}
+
+	if !atLeastOneValidAddress {
+		return nil, fmt.Errorf(
+			"addresses from modules is zero - "+
+				"check process permissions: '%s'", lastInvalidAddressLine)
```
Comment (fabled) [inline proc/proc.go:142]: Does the `lastInvalidAddressLine` give any additional information here? We could just say "all module addresses are zero" and drop the line numbers as irrelevant.

**Review by fabled — COMMENTED** [review]
Some stylistic comments. Could you also describe in the commit message or comments about what causes this? That is, when it is expected that one or few modules do not have valid address but others do?

File: proc/proc.go (hunk 2 above)
Comment (mchataigner) [reply proc/proc.go]: removed

File: proc/proc.go [hunk 3]
```diff
@@ -119,17 +122,21 @@ func GetKernelModules(modulesPath string,
...
 
 		symmap.Add(libpf.Symbol{
 			Name:    libpf.SymbolName(name),
 			Address: libpf.SymbolValue(address),
 			Size:    int(size),
 		})
 	}
+
+	if !atLeastOneValidAddress {
+		return nil, fmt.Errorf("addresses from all modules is zero - check process permissions")
```
Comment (rockdaboot) [inline proc/proc.go:137]:
```suggestion
		return nil, errors.New("addresses from all modules is zero - check process permissions")
```

**Comment by mchataigner** [comment]
Thanks for both reviews, I put some context inside the commit message and applied the suggestions.

File: proc/proc.go [hunk 4]
```diff
@@ -106,6 +106,9 @@ func GetKernelModules(modulesPath string,
 		Size:    int(etext.Address - stext.Address),
 	})
 
+	atLeastOneValidAddress := false
+	lastInvalidAddressLine := ""
```
Comment (rockdaboot) [inline proc/proc.go:110]: `lastInvalidAddressLine` isn't used any more. Did you run a `make lint` locally?
```suggestion
```

Comment (mchataigner) [reply proc/proc.go]: oups, sorry

**Comment by rockdaboot** [comment]
The integration tests obviously do not test the scan loop body. Is that expected or did this PR stumble on a flaw, @fabled?

Thanks for the fix, @mchataigner

**Comment by mchataigner** [comment]
I guess the integration tests were loading the kernel without any modules and my check was to have at least one kernel module with a non-null address.

I updated the condition to take this into account.

**Comment by fabled** [comment]
Could you rephrase the commit message to something like:
```
Improve module address check

Kernel v6.6 introduced the nls_ucs2_utils module. This module contains
only data without any code. Because of this, the module .text base is
always reported as 0x0.

The intent of the existing check for zero addresses is ensure correct
permissions for accessing processes and module metadata.

This commit allows having some kernel modules with address 0x0 and will
return an error only if all kernel modules are reporting address 0x0.
```

Otherwise looks good to me.

**Comment by mchataigner** [comment]
@fabled Done!
Thanks for the review.

**Review by fabled — APPROVED** [review]
:+1:
